### PR TITLE
GH-39327: [Java] define assemble descriptor for new custom maven plugin project

### LINKED ISSUE #39327: [Java][CI] java-jars CI job is failing on the maven plugin
### Describe the bug, including details regarding any error messages, version, and platform.

First failure: https://github.com/ursacomputing/crossbow/actions/runs/7192620687/job/19592544335

```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Apache Arrow Java Root POM 15.0.0-SNAPSHOT:
[INFO] 
[INFO] Arrow Maven Plugins ................................ FAILURE [ 22.208 s]
[INFO] Module Info Compiler Maven Plugin .................. SKIPPED
[INFO] Arrow Bill of Materials ............................ SKIPPED
[INFO] Apache Arrow Java Root POM ......................... SKIPPED
[INFO] Arrow Format ....................................... SKIPPED
[INFO] Arrow Memory ....................................... SKIPPED
[INFO] Arrow Memory - Core ................................ SKIPPED
[INFO] Arrow Memory - Unsafe .............................. SKIPPED
[INFO] Arrow Memory - Netty ............................... SKIPPED
[INFO] Arrow Vectors ...................................... SKIPPED
[INFO] Arrow Compression .................................. SKIPPED
[INFO] Arrow Tools ........................................ SKIPPED
[INFO] Arrow JDBC Adapter ................................. SKIPPED
[INFO] Arrow Flight ....................................... SKIPPED
[INFO] Arrow Flight Core .................................. SKIPPED
[INFO] Arrow Flight GRPC .................................. SKIPPED
[INFO] Arrow Flight SQL ................................... SKIPPED
[INFO] Arrow Flight SQL JDBC Driver Core .................. SKIPPED
[INFO] Arrow Flight SQL JDBC Driver ....................... SKIPPED
[INFO] Arrow Flight Integration Tests ..................... SKIPPED
[INFO] Arrow AVRO Adapter ................................. SKIPPED
[INFO] Arrow Algorithms ................................... SKIPPED
[INFO] Arrow Performance Benchmarks ....................... SKIPPED
[INFO] Arrow Java C Data Interface ........................ SKIPPED
[INFO] Arrow Orc Adapter .................................. SKIPPED
[INFO] Arrow Gandiva ...................................... SKIPPED
[INFO] Arrow Java Dataset ................................. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  23.631 s
[INFO] Finished at: 2023-12-13T10:09:09Z
[INFO] ------------------------------------------------------------------------
Error:  Failed to execute goal org.apache.maven.plugins:maven-assembly-plugin:3.6.0:single (default-cli) on project arrow-maven-plugins: Error reading assemblies: No assembly descriptors found. -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
Error: Process completed with exit code 1.
```

### Component(s)

Java

### PR DESCRIPTION
**Pull request by davisusanibar**
### Rationale for this change

To closes https://github.com/apache/arrow/issues/39327

### What changes are included in this PR?

GitHub CI validation needs to [run](https://github.com/apache/arrow/blob/main/ci/scripts/java_full_build.sh#L52) `assembly:single` for that reason is needed to setup a descriptor ref. In the case of this maven plugin, I only propose to include "src" as part of the resources.

### Are these changes tested?

Yes, by 
````
mvn clean \
    install \
    assembly:single \
    source:jar \
    javadoc:jar \
    -Papache-release \
    -DdescriptorId=source-release
````

### Are there any user-facing changes?

No.

* Closes: #39327

### CODE REVIEW

**Comment by davisusanibar** [comment]
@github-actions crossbow submit java

**Comment by davisusanibar** [comment]
@github-actions crossbow submit java-jars

File: java/maven/pom.xml [hunk 1]
```diff
@@ -281,6 +281,27 @@
           </execution>
         </executions>
       </plugin>
+      <!--
+      Generate custom Arrow Java Maven Plugin for module-info.java compiler.
+      Needed to pass Github CI validation `$ mvn assembly:single`
```
Comment (kou) [inline java/maven/pom.xml:286]:
```suggestion
      Needed to pass GitHub CI validation `$ mvn assembly:single`
```

**Comment by davisusanibar** [comment]
Could you please help me retry Build C++ libraries Ubuntu aarch64? @lidavidm / @kou 

Current [error](https://github.com/ursacomputing/crossbow/actions/runs/7281966542/job/19843469035):
````

The following tests FAILED:
	 53 - arrow-dataset-dataset-writer-test (Failed)
Error: `docker-compose --file /runner/_work/crossbow/crossbow/arrow/docker-compose.yml run --rm -e ARROW_JAVA_BUILD=OFF -e ARROW_JAVA_TEST=OFF java-jni-manylinux-2014` exited with a non-zero exit code 8, see the process log above.
````

**Comment by kou** [comment]
Retried!

**Comment by davisusanibar** [comment]
> Retried!

Well, there will be an issue with Build C++ libraries in Ubuntu aarch64. Is there any issue related to that?

**Comment by davisusanibar** [comment]
@github-actions crossbow submit java-jars

**Comment by davisusanibar** [comment]
> > Retried!
> 
> Well, there will be an issue with Build C++ libraries in Ubuntu aarch64. Is there any issue related to that?

Exclude failing tests to test the PR's scope

**Comment by kou** [comment]
> Well, there will be an issue with Build C++ libraries in Ubuntu aarch64. Is there any issue related to that?

https://github.com/apache/arrow/issues/39326

File: ci/scripts/java_jni_manylinux_build.sh [hunk 2]
```diff
@@ -115,7 +115,7 @@ if [ "${ARROW_BUILD_TESTS}" = "ON" ]; then
       # ImportError: ../grpc/_cython/cygrpc.cpython-38-aarch64-linux-gnu.so:
       # undefined symbol: vtable for std::__cxx11::basic_ostringstream<
       #   char, std::char_traits<char>, std::allocator<char> >
-      exclude_tests="${exclude_tests}|arrow-gcsfs-test"
+      exclude_tests="${exclude_tests}|arrow-gcsfs-test|arrow-dataset-dataset-writer-test"
```
Comment (lidavidm) [inline ci/scripts/java_jni_manylinux_build.sh:118]: nit: can you link the GitHub issue explaining why we are skipping this test?

**Review by lidavidm — APPROVED** [review]

File: ci/scripts/java_jni_manylinux_build.sh (hunk 2 above)
Comment (davisusanibar) [reply ci/scripts/java_jni_manylinux_build.sh:118]: It is a temporary workaround that only applies to testing this PR's scope. I think we should wait until https://github.com/apache/arrow/issues/39326 is resolved. Please let me know if this sound great for you, or, do we should push this PR with this exclude test to main branch?

Comment (lidavidm) [reply ci/scripts/java_jni_manylinux_build.sh:118]: Are we sure #39326 will be resolved in time?

**Comment by davisusanibar** [comment]
@github-actions crossbow submit java-jars

**Comment by davisusanibar** [comment]
@github-actions crossbow submit java-jar

**Comment by davisusanibar** [comment]
@github-actions crossbow submit java-jars

**Review by kou — APPROVED** [review]
+1

I can't review whether this approach is optimal or not but CI is green.

**Review by lidavidm — APPROVED** [review]

**Pull request merged**